File: frontend/lib/service/dictionary/DictionaryEntrySenseResource.ts
import axios from "axios";
import BACKENDROUTES from "../../BackendRoutes";

export function createDictionaryEntrySense(
    entryid: string,
    dname: string,
    uname: string,
    definition: string,
    order: number,
    get: Function = () => { }
) {
    const token = get('JWT') ?? '';

    const formData = new FormData();
    formData.append('entryid', entryid);
    formData.append('dname', dname);
    formData.append('uname', uname);
    formData.append('definition', definition);
    formData.append('order', order.toString());

    return axios.post(`${BACKENDROUTES.DICTIONARY}/entry/sense/create`, formData,
        {
            headers: {
                "Authorization": `Bearer ${token}`,
                'Content-Type': 'multipart/form-data'
            }
        }
    ).then(res => res.data);

}

export function updateDictionaryEntrySense(
    senseid: string,
    entryid: string,
    dname: string,
    uname: string,
    definition: string,
    order: number,
    get: Function = () => { }
) {
    const token = get('JWT') ?? '';

    const formData = new FormData();
    formData.append('senseid', senseid);
    formData.append('entryid', entryid);
    formData.append('dname', dname);
    formData.append('uname', uname);
    formData.append('definition', definition);
    formData.append('order', order.toString());

    return axios.post(`${BACKENDROUTES.DICTIONARY}/entry/sense/update`, formData,
        {
            headers: {
                "Authorization": `Bearer ${token}`,
                'Content-Type': 'multipart/form-data'
            }
        }
    ).then(res => res.data);
}

export function deleteDictionaryEntrySense(
    senseid: string,
    entryid: string,
    dname: string,
    uname: string,
    get: Function = () => { }
) {
    const token = get('JWT') ?? '';

    const formData = new FormData();
    formData.append('senseid', senseid);
    formData.append('entryid', entryid);
    formData.append('dname', dname);
    formData.append('uname', uname);

    return axios.post(`${BACKENDROUTES.DICTIONARY}/entry/sense/delete`, formData,
        {
            headers: {
                "Authorization": `Bearer ${token}`,
                'Content-Type': 'multipart/form-data'
            }
        }).then(res => res.data);
}

File: backend/src/main/java/de/garrafao/phitag/infrastructure/rest/dictionary/DictionaryEntrySenseResource.java
package de.garrafao.phitag.infrastructure.rest.dictionary;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import de.garrafao.phitag.application.dictionary.sense.DictionaryEntrySenseApplicationService;

@RestController
@RequestMapping(value = "/api/v1/dictionary/entry/sense")
public class DictionaryEntrySenseResource {

    private DictionaryEntrySenseApplicationService dictionaryEntrySenseApplicationService;

    @Autowired
    public DictionaryEntrySenseResource(DictionaryEntrySenseApplicationService dictionaryEntrySenseApplicationService) {
        this.dictionaryEntrySenseApplicationService = dictionaryEntrySenseApplicationService;
    }

    /**
     * Create a new sense for a dictionary entry.
     * 
     * @param authenticationToken The authentication token of the requesting user
     * @param entryid             The id of the dictionary entry
     * @param uname               The username of the dictionary owner (currently
     *                            will be ignored and replaced by the requesting
     *                            user)
     * @param dname               The name of the dictionary
     * 
     * @param definition          The definition of the sense
     * @param order               The order of the sense
     * 
     */
    @PostMapping("/create")
    public void create(
            @RequestHeader("Authorization") String authenticationToken,
            @RequestParam(value = "entryid") final String entryid,
            @RequestParam(value = "dname") final String dname,
            @RequestParam(value = "uname") final String uname,
            @RequestParam(value = "definition") final String definition,
            @RequestParam(value = "order") final Integer order) {
        dictionaryEntrySenseApplicationService.create(authenticationToken, entryid, dname, uname, definition, order);
    }

    /**
     * Update a sense for a dictionary entry.
     * 
     * @param authenticationToken The authentication token of the requesting user
     * @param senseid             The id of the sense
     * @param entryid             The id of the dictionary entry
     * @param uname               The username of the dictionary owner (currently
     *                            will be ignored and replaced by the requesting
     *                            user)
     * @param dname               The name of the dictionary
     * 
     * @param definition          The definition of the sense
     * @param order               The order of the sense
     */
    @PostMapping("/update")
    public void update(
            @RequestHeader("Authorization") String authenticationToken,
            @RequestParam(value = "senseid") final String senseid,
            @RequestParam(value = "entryid") final String entryid,
            @RequestParam(value = "dname") final String dname,
            @RequestParam(value = "uname") final String uname,
            @RequestParam(value = "definition") final String definition,
            @RequestParam(value = "order") final Integer order) {
        dictionaryEntrySenseApplicationService.update(authenticationToken, senseid, entryid, dname, uname, definition,
                order);
    }

    /**
     * Delete a sense for a dictionary entry.
     * 
     * @param authenticationToken The authentication token of the requesting user
     * @param senseid             The id of the sense
     * @param id                  The id of the dictionary entry
     * @param uname               The username of the dictionary owner (currently
     *                            will be ignored and replaced by the requesting
     *                            user)
     * @param dname               The name of the dictionary
     */
    @PostMapping("/delete")
    public void delete(
            @RequestHeader("Authorization") String authenticationToken,
            @RequestParam(value = "senseid") final String senseid,
            @RequestParam(value = "entryid") final String entryid,
            @RequestParam(value = "dname") final String dname,
            @RequestParam(value = "uname") final String uname) {
        dictionaryEntrySenseApplicationService.delete(authenticationToken, senseid, entryid, dname, uname);
    }

}
